### DISCUSSION TITLE
blur according to a mask

### DISCUSSION THREAD

**Discussion by mviereck** [ImageMagick/ImageMagick#5723] [Digital Image Processing] [answered]
I want to blur an image according to a mask.
The image should be blurred heavily where the mask is close to zero, and not at all where the mask is white. And respectively more or less where the mask is dark grey or light grey.
My task seems to be similar to `-adaptive-blur`.

(imagemagick 7.1.0-51 on Debian 11)

An example:
![maskmax](https://user-images.githubusercontent.com/14821372/200111217-837a7a88-df16-4b54-abbe-d09acdf331de.png)
![statistic](https://user-images.githubusercontent.com/14821372/200111221-41a96aa2-2364-4ed0-8f81-933e81966f82.png)

_(The example is created with my project [imfuse](https://github.com/mviereck/microscopy-tools), a focus stacking tool based on ImageMagick.)_

The image and its mask are generated from 243 images, making a focus stackshot to get an overall sharp image, where the single images are only partially sharp.

As you can see, the image looks well where the mask is bright, but has disturbing noise where the mask is dark.
My current solutions use thresholding or blurring the image everywhere where the mask is below 50%.
However, I desire a more advanced solution and think of an increasing/decreasing blur according to the mask values.

Do you have an idea how to achieve this?
I might try an iterative attempt where I threshold the mask to several grey levels and blur the according image parts in multiple steps.
But maybe a more direct solution is possible.

**Comment by snibgo** [comment, marked as answer]
There are many possibilities. For example, you can create a blurred version, and composite that over the original with a mask. However, that creates a ghosting effect. I generally prefer to use a mask that specifies the sigma, so we get true variable blur.

I explore these and other methods in [Selective blur](http://im.snibgo.com/selblur.htm).

**Reply by mviereck** [reply]
Thank you very much!
Checking your examples `-composite Blur` seems to do what I thought of. (I just need to `-negate` the mask above). 
I'll have a closer look at the other examples, too; maybe they provide even better solutions.

From https://www.imagemagick.org/Usage/mapping/#blur :
```
    magick image  blur_map \
        -define compose:args='{Xscale}[x{Yscale}[+{angle}]]' \
        -compose blur -composite   result
```
